src/TermStoreInstaller.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Wikibase\TermStore;
+
+interface TermStoreInstaller {
+
+	/**
+	 * @throws TermStoreException
+	 */
+	public function installOrUpdate();
```

Comment:
on naming this `installOrUpdate`.. this comes as a bit weird in an abstraction. like why assume there are two different behaviors `install` or `update` here. I would just call it `install` at this level

Reply:
I arrived at `installOrUpdate` from the user side, not from the implementer side. update.php is "update to latest version or install if it is not installed yet". So to me it seems like either we need a method like this, or expose more stuff to the caller can call `install` or `update`. Do you see another option?

Reply:
"update to latest version or install if it is not installed yet" sounds a bit concrete for this abstraction level. In databases, or things that persist long-term involving schemas, it might makes sense but that's not the case in other kinds of systems (file systems, key-value storages, in-memory caching mechanism ... etc).

so I would call this install, and if it means in some implementation that it creates or updates a table then let it be that way there.

Reply:
You are right there is not always an update. At the moment there is none for the Doctrine implementation, and indeed for some implementations you will never have one. The same can be said about install for stuff like in memory solutions.

The standard approach to interface design is to look at what the consumers need. Originally I figured there would be no sensible users when we where talking about if we need an interface like this at all. Turns out I was wrong and that we do have one. This consumer needs a way to tell the store to do any install and or update to get to its latest version. I see no problems from either a design or practical perspective if for some implementations this is a no-op.

I do OTOH see a problem with having an `install` method that also does updating. This is surprising. And from the interface you cannot tell this behavior is supported.

Reply:
> The standard approach to interface design is to look at what the consumers need. 

I'd like to stress this point. A lot of people do not have this in mind. Interfaces are about their consumers, not their implementers.

Reply:
> I do OTOH see a problem with having an install method that also does updating. This is surprising. And from the interface you cannot tell this behavior is supported.

but here you are still assuming that install and update are two distinct behaviors from a consumer that cares about the abstraction. If that is the case, then they should be separate.

>>    The standard approach to interface design is to look at what the consumers need.

> I'd like to stress this point. A lot of people do not have this in mind. Interfaces are about their consumers, not their implementers.

1. For that reason exactly I do think `installOrUpdate` is not thinking about all consumers. A consumer who wants to depend on this abstraction, does not want to worry about implementation details thus they do not even want to know what install means for whatever implementation is configured. So `install` is sufficient.

2. If you think about implementations on the other hand, which you also should do the way you do for consumers (and balance the too in a decent manner), giving me something that will be a no-op operation in my case is not desirable (that is not what abstraction is about anyway). Of course unless that's really necessary and it is a matter of edge case. I do not see here necessity nor edge-cases that will face this, rather a big chunk of implementers.

From 1 and 2 I conclude that the decent decision here is not to assume there are two different behaviors yet.

If you really disagree and think there's a good reason for distinction, then there should be a new concept in this abstraction, either a separate class `TermStoreUpdater` or a separate method `TermStoreInstaller::update()`.

To my mind, and I think it is a good interface design principle (also do not know if it has been stated by some other genius people, so I'll go ahead and call it Least Concepts Principle (LCP) :p ):
> one should not do half-way decisions in interfaces.. either a concept exist or not

----

Though, I do not want to strongly disagree, I am approving this for now and let it be an experiment of that principle (I'm happy to be proven wrong by receiving no complaints on the name in 20 years from now, well probably not worth complaining about in the first place, thus the approval ;) )

Reply:
For more clarity we should probably talk about this in person. Feel free to merge already is you deem that to be useful.